fix: pipeline run bugs in cyclic and acyclic pipelines

mathislucka opened the pull request:
### Related Issues

- fixes https://github.com/deepset-ai/haystack/issues/8657
- https://github.com/deepset-ai/haystack/issues/8641

### Proposed Changes:
The previous `Pipeline.run` logic [had several flaws](https://github.com/deepset-ai/haystack/pull/8695). Outputs of the pipeline depended on the insertion order of components or connections, cycles weren't processed in the right order, components ran more times than they should, and the output of a pipeline was not deterministic.

This PR introduces a new `Pipeline.run` logic that relies almost exclusively on the available data to decide when a component should run.

It is loosely based on the idea of [Kahn Process Networks](https://en.wikipedia.org/wiki/Kahn_process_networks) (a distributed model of computing).

**Core Logic**
Internally, the `Pipeline.run` method uses a single priority queue to schedule components to run. Components are popped from the priority queue one at a time until there are no more components that can run. Once there are no more components that can run, the pipeline returns an output. The queue is re-computed when needed because the produced outputs of the components in the pipeline might change priorities. Since pipelines can have cycles, the component can enter the priority queue multiple times.


**When can a component run?**
A component can run once it passes two conditions:
1. All mandatory inputs are available (i.e. if we call `component.run` we do not get an exception because of missing positional or keyword arguments).
2. The component was triggered to run.


**Triggers**
There are three types of triggers that cause a component to run:
1. The component receives input from outside the pipeline (e.g. user input)
2. The component receives input from another component in the pipeline
3. The component does not have incoming connections to any other component in the pipeline and `Pipeline.run` is called

A trigger is "consumed" by the component, meaning that it can only cause it to run once. For example, for each `Pipeline.run` invocation, the component can only run once because of user input. It could still run again, but it needs to receive a second trigger to do that.

A component does not always run immediately when it receives a trigger, it only runs when it has highest priority in the priority queue.


**Priorities**
At a high level, we differentiate between components that can't run, because they don't fulfil the conditions to run.  Components that can run immediately in any order, and components that we want to run later because they might still receive optional or lazy variadic inputs.


**Inputs & Outputs**
A component can receive inputs and it can produce outputs.

When a component runs, it "consumes" (deletes) its inputs, meaning that these same inputs will not be available in case the component runs another time. Inputs from outside the pipeline are an exception to this rule. They are only consumed when they are passed to a `GreedyVariadic` socket of the component (e.g. `BranchJoiner`). Other inputs from outside the pipeline will always be available to a component, no matter how often it runs.

After a component ran, its outputs are distributed to the connected input sockets of other components in the pipeline. Outputs that are not connected to any other input sockets in the pipeline, are returned to the user.


**Impact on existing pipelines**

**Non-Cyclic Pipelines**
For non-cyclic pipelines, the execution order of components might change. This does not have any impact on the outputs of a pipeline, except in one condition:

If the pipeline has two branches that are joined by a lazy variadic socket (e.g. `DocumentJoiner`), the order of the joined inputs might change. In the existing pipeline logic, the order is determined by the order of adding components to the pipeline and the order of connecting components. This behavior is not documented anywhere and the user can't know in which order these components will be executed without studying the underlying implementation in Haystack and NetworkX. This PR introduces a lexicographical sort for these cases, other possibilities could be discussed. We can potentially provide the users with utility functions to test if the output might change, when we release the changes from this PR.

**Cyclic Pipelines**
Cyclic pipelines are affected by several bugs in the current pipeline logic.
Cyclic pipelines might be affected if they meet any of these conditions:
- pipelines with more than one optional or greedy variadic edge in the cycle (e.g. PromptBuilder, BranchJoiner)
- pipelines with two cycles that share an optional or greedy variadic edge
- ... (more?)

For these conditions, neither the run order nor the outputs of the pipeline might be deterministic (i.e. the output of the pipeline might change although the code didn't change). Again, it should be possible to provide tooling that helps users understand if their pipelines are affected.

**Open Issues**

**Cycles without a defined entrypoint**
For cycles with more than one component that only have default inputs and that receive inputs from outside the pipeline, there is no defined entrypoint (i.e. we can not know which component in the cycle should run first).

Consider this pipeline:
![image](https://github.com/user-attachments/assets/23b66ddc-b0a8-448d-8377-5de2f444fb51)


The use case is that one LLM generates code that is checked by another LLM for correctness and then either returned to the first LLM if it has feedback or returned to the user, if the second LLM decides that the code is good enough. The instructions for the "Feedback LLM" and the task for the "Code LLM" are both provided by the user from outside the cycle. Since we have 2 `PromptBuilder` components in the cycle (configured so that all inputs are optional) and they both receive an input that triggers them to run at the same time, and they are both waiting for exactly one more input (but can run without it), there is no defined order for which of these components should be executed first.

At least from my understanding, this problem can't be solved purely based on the topology of the pipeline graph or on the available data.

In this implementation, the `code_prompt` would run first because `c` comes before `f` in a lexicographical sort. We can document this behavior but it might still surprise our users. Additional measures could be:
- test for this condition and log a warning
- test for this condition and raise an exception (the user could make the edge from `code_llm` to `feedback_prompt` required to solve the problem)
- give users another way to specify the running order for components that could run at the same time when the running order could affect the outputs of the pipeline (might be complex)

**Should cycles run like loops in a programming language?**

@tstadel pointed out that components outside of a cycle might receive inputs from the cycle while the cycle is still running. In very few edge cases, this could cause the components outside the cycle to run repeatedly (and in turn triggering other components in the pipeline).

When we think of our pipeline in terms of a distributed model of computing, then this behavior would be expected.

However, if we assume that a cycle in a pipeline works the same way as a loop in a programming language, then the loop should run to completion before we execute any other components.

Consider this pseudo-implementation:
```python
outside_output = None
for component in cycle:
  if component.output_receiver == 'outside_output':
    outside_output = component.output

# continue with the last value that was set for outside_output
```

Another option would be to treat the input socket of a component like a FIFO queue, meaning that the outputs would not be overwritten in case components in the cycle provide the output multiple times.
```python
outside_output = []
for component in cycle:
  if component.output_receiver == 'outside_output':
    outside_output.append(component.output)

# receiving component will run as many times as outputs were appended to outside_output
```

My recommendation would be to follow the distributed model of computing approach and allow a component to run as soon as it receives inputs and is ready to run. The implementation is a lot less complex, especially so, when we introduce concurrency to our pipeline execution. If the user does not want the component outside the cycle to run before the cycle has fully executed, it is easy to achieve that by a different dataflow in the pipeline or by marking edges as required.

### How did you test it?
- behavioral tests
- some tests that are currently failing are expected to fail, because we are testing the running order of the components, which is not actually the behavior that we want to test

### Notes for the reviewer
This is a work in progress.

The change needs to be tested extensively.

My recommendation would be to update the behavioral tests so that we test for the inputs that a component received to run and the pipeline outputs instead of testing for the component execution order.

Using content tracing, we can adapt the existing testing approach by exchanging `expected_run_order` with `expected_component_inputs`. Expected component inputs could be tested like this (pseudo-code):

```python
expected_component_inputs = {('<component_name>', '<visits>'): {...inputs}}

for key, inputs in actual_inputs.items():
   assert inputs == expected_component_inputs.get(key)
```

This is better than the current approach because we don't really want to test the execution order of the components, we only care about how often a component runs, if it has the same inputs, and if the pipeline has the same outputs.

Currently, the behavioral tests do not test these two behaviors ([as demonstrated here](https://github.com/deepset-ai/haystack/pull/8695)):
- how often a component runs
- if it has the same inputs

For real-world use cases, changes in these two behaviors **will** have an impact on the output of the pipeline.

Adapting the tests allows us to re-use the test suite when we implement pipelines that run components concurrently.


### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:` and added `!` in case the PR includes breaking changes.
- I documented my code
- I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

mathislucka commented on test/core/pipeline/features/test_run.py:
```diff
@@ -2821,3 +2821,148 @@ def run(self, replies: List[str]):
     )
+
+@given("a pipeline that passes outputs that are consumed in cycle to outside the cycle", target_fixture="pipeline_data")
+def passes_outputs_outside_cycle():
```
@tstadel could you check if this test covers the use case that you had with passing the prompt outside the cycle?

mathislucka replied:
The test fails on main and it yields an empty prompt instead of the expected prompt value.

mathislucka replied:
The output from the cycle isn't distributed in the old logic because we are filtering for components that are in the cycle here:

https://github.com/deepset-ai/haystack/blob/167ede1f4ca8996aa3fece336ba13fde85724e95/haystack/core/pipeline/pipeline.py#L213

And then we remove any output that was already received by a component in the cycle here: https://github.com/deepset-ai/haystack/blob/167ede1f4ca8996aa3fece336ba13fde85724e95/haystack/core/pipeline/base.py#L889

That means that outputs that are received both by components inside and outside the cycle will only go to components inside the cycle.

mathislucka commented on haystack/core/pipeline/pipeline.py:
```diff
@@ -40,6 +40,72 @@ class Pipeline(PipelineBase):
     Orchestrates component execution according to the execution graph, one after the other.
     """
 
+    def _warn_if_ambiguous_intent(
```
@Amnah199 I started out with this method to check for ambiguous execution order up front (e.g. the case that I showed for the cycle with 2 prompt builders in the PR description).

The problem is, that we don't know for sure that the running order will be ambiguous, we just know that it's possible.

I then thought that we might just check this when actually running the pipeline and only warn if this case occurs (see comment further down).

I think that would be a better approach and I'd skip this one then. What do you think?

mathislucka commented on haystack/core/pipeline/pipeline.py:
```diff
@@ -362,191 +491,63 @@ def run(  # noqa: PLR0915, PLR0912
+                    next_priority, next_name = priority_queue.peek()
+
+                # alternative to _warn_if_ambiguous_intent
+                if priority in [ComponentPriority.DEFER, ComponentPriority.DEFER_LAST] and next_priority == priority:
```
Alternative check that would only warn if ambiguity actually occurs.

Amnah199 replied:
Based on my current understanding, I prefer this approach; however, let’s keep both options for now before making a final decision.

I do have one question: why are we referring to this as an "ambiguity"? In this case, the component that comes first in lexicographical order should execute. Since we also plan to document this for users to clarify the execution order, we can simply notify them that, based on the sorting criteria, if we have lets say a `retriever` and a `summarizer` waiting in this particular scenario, then `retriever` will run first. `"Ambiguous running order"` suggests that we expect the order to unpredictably change in certain cases, which I believe shouldn’t happen. Please let me know if I’ve misunderstood anything.

mathislucka replied:
Yeah, I wasn't sure about the wording here.

The execution order is clear, the component that comes first lexicographically will also execute first. I just thought that this is a pretty implicit mechanism and the user might not be aware of it (many don't read the docs). So this warning is there to pull the users attention to the fact that the same priority is resolved by using lexicographical sorting to check if it was also the user's "intent" to run the components in that order.

Amnah199 commented on haystack/core/pipeline/component_checks.py:
```diff
@@ -0,0 +1,244 @@
+    for socket_name, socket in sockets_to_check.items():
+        socket_inputs = inputs.get(socket_name, [])
+        expected_sockets.add(socket_name)
+        if (
+            is_socket_lazy_variadic(socket)
+            and any_socket_input_received(socket_inputs)
+            or has_socket_received_all_inputs(socket, socket_inputs)
+        ):
```
A minor suggestion for code efficiency, and this might make it easier to understand the intended logic:
```suggestion
        # Check if socket has all required inputs or is a lazy variadic socket with any input

       if (
    has_socket_received_all_inputs(socket, socket_inputs)
    or (is_socket_lazy_variadic(socket) and any_socket_input_received(socket_inputs))
):
```

Amnah199 commented on test/core/pipeline/features/test_run.py:
```diff
@@ -1523,10 +1525,36 @@ def run(self, query_embedding: List[float]):
+                        ],
+                        "top_k": None,
+                    },
+                },
```
@mathislucka I am adding these sort of checks with content tracing for component calls. Can you verify if this aligns with what you described?

mathislucka replied:
Yes, that's exactly what I had in mind. 🙌 

They would replace the run_order tests but we can add them first and then remove run orders in one go.

Amnah199 commented on haystack/core/pipeline/pipeline.py:
```diff
@@ -71,191 +172,220 @@ def _run_component(
+        """
+        if not can_component_run(component, inputs):
+            return ComponentPriority.BLOCKED
+        elif is_any_greedy_socket_ready(component, inputs) and are_all_sockets_ready(component, inputs):
+            return ComponentPriority.HIGHEST
```
@mathislucka, could you help clarify our goal behind assigning “highest priority” to components?

If `HIGHEST` means running as soon as all mandatory sockets are ready and at least one greedy socket has inputs, then we should consider setting `only_check_mandatory=True`. I thought about this because as per my understanding a “greedy” socket implies the component is willing to run multiple times or run as soon as partial input is available. Is the goal that we don’t block execution due to optional sockets that haven’t yet received inputs?
```suggestion
        elif is_any_greedy_socket_ready(component, inputs) and are_all_sockets_ready(component, inputs, only_check_mandatory=True):
            return ComponentPriority.HIGHEST
```

mathislucka replied:
Yes, I initially did not wait for optional inputs but I checked the existing logic and we wait for optional inputs there, so I did not want to change that.

I think that in practice we don't have any components that receive both a greedy variadic and an optional input.

Amnah199 commented on haystack/core/pipeline/pipeline.py:
```diff
@@ -2,56 +2,157 @@
+        # Pipeline has no components.
+        if len(pq) == 0:
+            return
+
```
@mathislucka, currently [this test ](https://github.com/deepset-ai/haystack/blob/f96839e139c3e4395a847481aa4512343508e602/test/core/pipeline/features/test_run.py#L1932)is failing as an empty list is returned instead of `PipelineRuntimeError`. The pipeline has no valid entry point and is not executed at all so we should raise an error.
To address this, we could add a small validation check before entering the execution loop. Below is a suggestion—what do you think?

```suggestion
            # check if pipeline is blocked before running
            self.validate_pipeline(priority_queue)
            
         # function definitions
     @staticmethod
    def _is_queue_blocked(pq: FIFOPriorityQueue) -> bool:
        queue_copy = deepcopy(pq)
        while queue_copy:
            component = queue_copy.peek()
            if component[0] != ComponentPriority.BLOCKED:
                return False
            queue_copy.pop()
        return True
        
    def validate_pipeline(self, priority_queue: FIFOPriorityQueue):
        """
        Validate the pipeline to check if it is blocked or has no entry point
        :param priority_queue: Priority queue of component names.
        """
        if self._is_queue_blocked(priority_queue):
            raise PipelineRuntimeError("Cannot run pipeline - all components are blocked. This typically happens when:\n"
            "1. Components are missing required inputs\n"
            "2. There is a circular dependency without a valid entry point\n"
            "Check the connections between these components and ensure all required inputs are provided.")
```

mathislucka commented:
@Amnah199 @julian-risch @davidsbatista

Updated the PR to the latest state from experimental.
It also incorporates the changes that are needed for the AsyncPipeline (already merged to experimental).
I'll raise the AsyncPipeline-change as a separate PR.

Here is what was added in addition to the latest changes:
- added a test for thread-safe execution
- added a test for the file conversion case that we discussed
- added missing unit tests for the FIFOPriorityQueue
- completed the work from @Amnah199 on testing for component inputs instead of run order
- added a release note

Amnah199 commented on test/core/pipeline/features/test_run.py:
```diff
@@ -1333,14 +1683,104 @@ def run(self, prompt: str):
-                    "llm",
-                    "output_validator",
-                ],
+                expected_component_calls={
```
@mathislucka I was considering the option of replacing these few very long component_calls with `<Any>`. If think they are convoluting the file.

Amnah199 replied:
Sure, we can keep them now. I'll create a separate issue to investigate how to refactor/optimize these tests for better testing and maintainability over time.

davidsbatista commented on haystack/core/pipeline/utils.py:
```diff
@@ -18,3 +20,103 @@ def parse_connect_string(connection: str) -> Tuple[str, Optional[str]]:
+        :param priority:
+            Priority level for the item. Lower numbers indicate higher priority.
+        """
+        count = next(self._counter)
```
although within a different scope, this variable `count` has the same name as `from itertools import count`
maybe a different name like `order` - just to be safe

davidsbatista commented:
@mathislucka can you check your `base.py` and the new work done with mermaid ink that was recently merged? your PR if merged would delete it.

mathislucka commented:
> @mathislucka can you check your `base.py` and the new work done with mermaid ink that was recently merged? your PR if merged would delete it.

I'll check, the PR was merged in between me merging `main` to this branch and changing `base`. I'll check all commits since the last merge and re-apply the changes.

mathislucka commented:
@davidsbatista thanks for catching this. I should have been more careful when copying over the changes from experimental.

davidsbatista commented on haystack/core/pipeline/pipeline.py:
```diff
@@ -349,204 +185,81 @@ def run(  # noqa: PLR0915, PLR0912
-        # Reset the visits count for each component
-        self._init_graph()
-
         # TODO: Remove this warmup once we can check reliably whether a component has been warmed up or not
```
As far as I've understood, this is to avoid a component being "warmed up" several times, which can happen if a pipeline is run multiple times - If it's not already handled, maybe it's worth opening an issue for this.

mathislucka replied:
No, this only reset the component visits to 0. The warm_up is handled in the `warm_up`-method (https://github.com/deepset-ai/haystack/blob/82d95e0bc75f6f0c623a43c36957c37f002fc4d2/haystack/core/pipeline/base.py#L725).

We can remove resetting the visits to 0 because we are not tracking visits as instance state anymore. Instead, we track visits on a per-run invocation basis.

Previously, pipeline.run was not thread-safe because the visits were tracked on the instance level, so one run invocation would impact the other.

davidsbatista replied:
sorry, my comment is related to the `self.warm_up()` below - I just started writing it in association with the `# TODO` line - we should check ( open an issue to work on this) if a component was warmed up already, before doing so, since this can delay a pipeline run.

davidsbatista replied:
It's not a blocker - I just wanted to have a reminder/issue so that we can later work on this.

We can agree on how to handle those. For instance, every component that has a `warm_up()` method should also have a `self.warmed_up : bool = True|False`

davidsbatista replied:
@mathislucka I will open an issue to track this

mathislucka replied:
https://github.com/deepset-ai/haystack/issues/8823

davidsbatista commented on pyproject.toml:
```diff
@@ -185,7 +185,7 @@ include = ["/haystack", "/VERSION.txt"]
 packages = ["haystack"]
 
 [tool.codespell]
-ignore-words-list = "ans,astroid,nd,ned,nin,ue,rouge,ist"
```
I'm just curious: where did the "Claus" come from?

Amnah199 replied:
It was raising a linting error in one of the PRs. The linter was forcing to rename "Santa Claus" to "Santa Clause".

davidsbatista commented on test/conftest.py:
```diff
@@ -4,14 +4,15 @@
 from datetime import datetime
```
this import can be removed, not being used

davidsbatista commented on test/conftest.py:
```diff
@@ -4,14 +4,15 @@
 from test.tracing.utils import SpyingTracer
-from typing import Generator
+from typing import Generator, Dict
 from unittest.mock import Mock, patch
```
`patch` can also be removed, not being used

davidsbatista commented on test/core/pipeline/test_component_checks.py:
```diff
@@ -0,0 +1,645 @@
+import pytest
+from typing import Dict, Any
```
these imports are also not being used

davidsbatista commented on test/core/pipeline/test_pipeline.py:
```diff
@@ -1,53 +1,17 @@
-from typing import List, Optional
-from unittest.mock import patch
+
+from typing import Optional
```
these can be removed

davidsbatista commented on test/core/pipeline/test_pipeline.py:
```diff
@@ -1,53 +1,17 @@
-    @component.output_types(value=str)
-    def run(self, input_: str):
-        return {"value": input_}
+from haystack.core.component.types import InputSocket, OutputSocket, Variadic
```
can also be removed

davidsbatista commented:
This was already a second round of review after this went into the experimental. Unfortunately, we did not receive much feedback from the experimental release.

A few details to consider
- after merging we should also open an issue to remove this new Pipeline code from `haystack-experimental`
- not a priority - but, we should rename the `utils.py` to something more meaningful (e.g.: priority_queue.py) since 90% of the file content is related to the FIFO priority queue

davidsbatista reviewed (approved):
LGTM - this was the second review (first was when it was merged in experimental) - left a few comments related unused imports and some details about next steps. 

Thaks for taking this over @mathislucka 👍🏽

Amnah199 commented on haystack/core/pipeline/component_checks.py:
```diff
@@ -0,0 +1,245 @@
+    :param socket_inputs: Inputs for the socket.
+    """
+    # No inputs received for the socket, it is not filled.
+    if len(socket_inputs) == 0:
```
Minor improvement
```suggestion
    if not socket_inputs:
```

mathislucka replied:
I like checks to be explicit. It's easy to get tripped up with different types of falsy values. This tests explicitly for what I'm expecting. Do we have a convention for this in Haystack?

Amnah199 replied:
We don't have a strict convention, its just more pythonic to use `if not socket_inputs:`.
But its up to you, if you think the explicit check is more safe.

mathislucka replied:
I'd keep it this way.
In my view, the benefits are:
- safer against accidental changes
- immediately tells devs reading this code that socket_inputs is an `Iterable`

Generally speaking, I wouldn't say that the implicit evaluation of falsy/truthy values is more pythonic. This is what Python core developer Alyssa Coghlan has to say about `if x:` vs. `if x is not None`:

> When using None as a sentinel for such types, writing
"if x:" when you really mean "if x is not None:" is a harmless style
error, with no practical ill-effects.

(https://mail.python.org/pipermail/python-ideas/2014-March/026647.html)

Overall, I think it comes down to taste or convention.

Amnah199 commented on haystack/core/pipeline/pipeline.py:
```diff
@@ -71,191 +75,21 @@ def _run_component(
-
-            if not isinstance(res, Mapping):
+            span.set_content_tag("haystack.component.input", deepcopy(component_inputs))
+            logger.info("Running component {component_name}", component_name=component_name)
```
```suggestion
            logger.info(f"Running component {component_name}")
```

mathislucka replied:
https://docs.astral.sh/ruff/rules/logging-f-string/ -> we should not use f-strings in logging. We even had a case in Haystack v1 where removing f-strings from logging brought major memory reduction and performance improvements: https://github.com/deepset-ai/haystack/pull/3212

davidsbatista commented:
@mathislucka I forgot to mention, I think we should elaborate a bit more the release notes, this is a big change - maybe mention it was refactored, new logic introduced - before saying that certain cases/bugs were fixed.

Amnah199 commented on releasenotes/notes/fix-pipeline-run-2fefeafc705a6d91.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+fixes:
+  - |
+    Fixes multiple bugs in the pipeline-run-method:
+    - acyclic pipelines with multiple lazy variadic components not running all components
```
I think we can add here something like:

"`Pipeline.run` logic now uses data drive component execution rather than graph based cycle handling"

Amnah199 reviewed (changes requested):
The PR already looks solid. Just last few updates left. Then we can go ahead and merge!

Amnah199 commented on releasenotes/notes/fix-pipeline-run-2fefeafc705a6d91.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+fixes:
+  - |
+    Major refactoring of Pipeline.run to fix multiple bugs.
+    We moved from a mostly graph-based to a dynamic dataflow driven execution logic.
+    Check your pipeline executions carefully to see if their output changes.
+    Bugs fixed:
+    - acyclic pipelines with multiple lazy variadic components not running all components
+    - cyclic pipelines not passing intermediate outputs to components outside the cycle
+    - cyclic pipelines with two or more optional or greedy variadic edges showing unexpected execution behavior
+    - cyclic pipelines with two cycles sharing an edge raising errors
```
```suggestion
---
highlights: >
  Major refactoring of `Pipeline.run()` to fix multiple bugs.
    We moved from a mostly graph-based to a dynamic dataflow driven execution logic.
    Check your pipeline executions carefully to see if their output changes.
fixes:
- |
    - acyclic pipelines with multiple lazy variadic components not running all components
    - cyclic pipelines not passing intermediate outputs to components outside the cycle
    - cyclic pipelines with two or more optional or greedy variadic edges showing unexpected execution behavior
    - cyclic pipelines with two cycles sharing an edge raising errors
```

Amnah199 reviewed (approved):
Great work! Thanks for your contribution @mathislucka.

mathislucka commented:
Thanks for bearing with me on this one @Amnah199 and @davidsbatista! This was a tough review 🙌
